openverse_catalog/dags/providers/provider_dag_factory.py

```diff
@@ -93,105 +93,83 @@
-                       which data should be ingested).
-    day_shift:         integer giving the number of days before the
-                       current execution date the `main_function` should
-                       be run (if `dated=True`).
-    pull_timeout:      datetime.timedelta giving the amount of time a given data
-                       pull may take.
-    load_timeout:      datetime.timedelta giving the amount of time a given load_data
-                       task may take.
-    doc_md:            string which should be used for the DAG's documentation markdown
-    media_types:       list describing the media type(s) that this provider handles
-                       (e.g. `["audio"]`, `["image", "audio"]`, etc.)
-    create_preingestion_tasks and create_postingestion_tasks: callable which creates a
-                        task or task group to be run before or after (respectively) the
-                        rest of the provider workflow. Loading and dropping temporary
-                        tables is one example.
+    conf: ProviderWorkflow configuration object.
     """
-    default_args = {**DAG_DEFAULT_ARGS, **(default_args or {})}
-    media_type_name = "mixed" if len(media_types) > 1 else media_types[0]
-    provider_name = dag_id.replace("_workflow", "")
-    identifier = f"{provider_name}_{{{{ ts_nodash }}}}"
+    default_args = {**DAG_DEFAULT_ARGS, **(conf.default_args or {})}
 
     dag = DAG(
-        dag_id=dag_id,
-        default_args={**default_args, "start_date": start_date},
-        max_active_tasks=max_active_tasks,
-        max_active_runs=max_active_runs,
-        start_date=start_date,
-        schedule_interval=schedule_string,
-        catchup=dated,  # catchup is turned on for dated DAGs to allow backfilling
-        doc_md=doc_md,
-        tags=["provider"] + [f"provider: {media_type}" for media_type in media_types],
+        dag_id=conf.dag_id,
+        default_args={**default_args, "start_date": conf.start_date},
+        max_active_tasks=conf.max_active_tasks,
+        max_active_runs=conf.max_active_runs,
+        start_date=conf.start_date,
+        schedule_interval=conf.schedule_string,
+        catchup=conf.dated,  # catchup is turned on for dated DAGs to allow backfilling
+        doc_md=conf.doc_md,
+        tags=["provider"]
+        + [f"provider: {media_type}" for media_type in conf.media_types],
         render_template_as_native_obj=True,
         user_defined_macros={"date_partition_for_prefix": date_partition_for_prefix},
     )
 
     with dag:
+        ingest_data, ingestion_metrics = create_ingestion_workflow(conf)
+
+        report_load_completion = create_report_load_completion(
+            conf.dag_id, conf.media_types, ingestion_metrics, conf.dated
+        )
+
+        ingest_data >> report_load_completion
+
+    return dag
+
+
+def create_ingestion_workflow(conf: ProviderWorkflow, day_shift: int = 0):
```

Comment:
For sure, sorry -- I tend to prefer the opposite order, but I think your view is the more popular and technically more 'correct' one 😆 I just find it easier to read the 'top-level' function first, rather than scrolling past the definition of functions whose context has not yet been established and then having to go back.

I do think it should remain in this file; if anything it's the bulk of the factory work, not really a utility.

Reply:
I totally get that and have heard the same from others! I just happen to find myself going through that top level function and have no idea or context on what the functions called within do which confuses me more 😅 Thanks for being willing to move it around though!